src/core/Track.cpp

```diff
@@ -609,6 +746,21 @@ void TrackContentObjectView::dropEvent( QDropEvent * de )
 	QDomElement tcos = dataFile.content().firstChildElement( "tcos" );
 	m_tco->restoreState( tcos.firstChildElement().firstChildElement() );
 	m_tco->movePosition( pos );
+	auto old_tco = dynamic_cast<TrackContentObjectView *>( qwSource )->m_tco;
```

Comment:
Is it always the case that `qwSource` is an instance of `TrackContentObjectView`? If so, use `static_cast` instead. If not, ensure that `dynamic_cast` doesn't result in `nullptr` before accessing the result. (This should always be considered when using `dynamic_cast` with pointer types.)

Reply:
A dynamic cast is used some five or six lines above to check whether it's pasting the same TCO into itself. I'm going to stick to the dynamic cast to stay safe, but I'll add a check for a `nullptr`.

Reply:
Sorry, I didn't explain the problem very well. `dynamic_cast` checks at runtime whether its argument really is an instance of the type you asked it to cast to. If it isn't, and the target type is a pointer type, the result is `nullptr`. (If it isn't, and the target type is a reference type, it throws `std::bad_cast`.)

There are two possibilities:
* The argument is always an instance of the target type. In this case, you don't need the extra overhead of `dynamic_cast`'s runtime checks, so use `static_cast` instead.
* The argument isn't always an instance of the target type. In this case, `dynamic_cast` may result in `nullptr`, even if its parameter is never null. You should check whether the result is `nullptr` before you try to dereference it (in this case, access `->m_tco`). It doesn't make sense to ask `dynamic_cast` to check whether the cast is successful or not, only to ignore it and dereference the result anyway.

The `dynamic_cast` above is fine, because the result is never dereferenced. It's just compared to `this`, so if the result is `nullptr`, the comparison fails, which is the intended behaviour.

I would write something like
```cpp
const auto tcovSource = dynamic_cast<TrackContentObjectView *>(qwSource);
if (tcovSource && tcovSource->m_tco->usesCustomClipColor())
{
    // ...
```

You could declare `tcovSource` a little further up and reuse it in place of both `dynamic_cast`s, which would be a little less repetitive and a little more efficient.

Reply:
I've now done what you've shown in the code. I decided to leave the other one as is as I think it's a check that has to be done, and the second is an initialisation, and they are also handled differently on true/false scenarios. You can resolve this if you feel this is satisfactory.